model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java

```diff
@@ -415,15 +415,51 @@ public void moveGroup(RealmModel realm, GroupModel group, GroupModel toParent) {
         group.setParent(toParent);
         if (toParent != null) toParent.addChild(group);
         else session.groups().addTopLevelGroup(realm, group);
+
+        // TODO: Remove em.flush(), currently this needs to be there to translate ConstraintViolationException to
+        //  DuplicateModelException {@link PersistenceExceptionConverter} is not called if the
+        //  ConstraintViolationException is not thrown in method called directly from EntityManager
+        em.flush();
     }
 
     @Override
     public Stream<GroupModel> getGroupsStream(RealmModel realm) {
-        RealmEntity ref = em.getReference(RealmEntity.class, realm.getId());
+        return closing(em.createNamedQuery("getGroupIdsByRealm", String.class)
+                .setParameter("realm", realm.getId())
+                .getResultStream())
+                .map(g -> session.groups().getGroupById(realm, g));
+    }
+
+    @Override
+    public Stream<GroupModel> getGroupsStream(RealmModel realm, Stream<String> ids, String search, Integer first, Integer max) {
+        if (search == null) search = "";
+
+        TypedQuery<String> query = em.createNamedQuery("getGroupIdsByNameContainingFromIdList", String.class)
+                .setParameter("realm", realm.getId())
+                .setParameter("search", search)
+                .setParameter("ids", ids.collect(Collectors.toList()));
+
+        if (first != null && first > 0) {
+            query = query.setFirstResult(first);
+        }
+
+        if (max != null && max > 0) {
+            query = query.setMaxResults(max);
+        }
+
+        return closing(query.getResultStream())
+                .map(g -> session.groups().getGroupById(realm, g));
+    }
 
-        return ref.getGroups().stream()
-                .map(g -> session.groups().getGroupById(realm, g.getId()))
-                .sorted(Comparator.comparing(GroupModel::getName));
+    @Override
+    public Long getGroupsCount(RealmModel realm, Stream<String> ids, String search) {
+        if (search == null) search = "";
+
+        return em.createNamedQuery("getGroupCountByNameContainingFromIdList", Long.class)
```

Comment:
same as above

Reply:
I believe there is no need for such a function. We should probably throw an exception here if the search is `null`. It doesn't make sense to search for a count of groups with ids (we can just use the size of ids list). In the adapter layer, we should make sure no such call is made.

Reply:
Looked at it again and an exception is probably not the best, but we can just call `ids.map(id -> session.groups().getGroupById(realm, id))`, it should be probably faster, WDYT?

Reply:
I see, ok I fixed